components/locale_core/src/langid.rs

```diff
@@ -183,9 +185,46 @@ impl LanguageIdentifier {
     ///     Ok("pl-Latn-PL")
     /// );
     /// ```
-    pub fn canonicalize<S: AsRef<[u8]>>(input: S) -> Result<String, ParseError> {
+    pub fn canonicalize_utf8(input: &[u8]) -> Result<Cow<str>, ParseError> {
         let lang_id = Self::try_from_utf8(input.as_ref())?;
-        Ok(lang_id.write_to_string().into_owned())
+        let cow = lang_id.write_to_string();
+        if cow.as_bytes() == input {
+            if let Ok(s) = core::str::from_utf8(input) {
+                Ok(s.into())
+            } else {
+                Ok(cow.into_owned().into())
```

Comment:
this is unreachable. `try_from_utf8` succeeding should allow you to justify `from_utf8_unchecked`